community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/PropertyContainerStateImpl.java

```diff
@@ -222,18 +211,33 @@ public boolean hasChanges()
     @Override
     public StorageProperty getChangedProperty( int propertyKeyId )
     {
-        return changedProperties != null ? changedProperties.get( propertyKeyId ) : null;
+        return changedProperties == null ? null : getPropertyOrNull( changedProperties, propertyKeyId );
     }
 
     @Override
     public StorageProperty getAddedProperty( int propertyKeyId )
     {
-        return addedProperties != null ? addedProperties.get( propertyKeyId ) : null;
+        return addedProperties == null ? null : getPropertyOrNull( addedProperties, propertyKeyId );
     }
 
     @Override
     public boolean isPropertyRemoved( int propertyKeyId )
     {
         return removedProperties != null && removedProperties.containsKey( propertyKeyId );
     }
+
+    private Iterator<StorageProperty> toPropertyIterator( VersionedHashMap<Integer,Value> propertyMap )
+    {
+        return propertyMap == null ? Collections.emptyIterator() :
+                Iterators.map(
+                    entry -> new PropertyKeyValue( entry.getKey(), entry.getValue() ),
```

Comment:
Yes it does. Honestly the whole `PropertyKeyValue` is a little subpar considering the hotness of that path. However, we are about to go over cursors anyway, so I didn't want to over work this now. Introduced `PropertyKeyValue` only to make the current cursors and iterators work as before.

I suggest we put reducing object allocation during property iteration into a separate card and revisit later in 3.3 when we have some cursors in place. Then it might already be fixed.